File: svc-java/src/main/java/com/hp/vvs/app/alias/AliasRegistry.java
package com.hp.vvs.app.alias;

import com.hp.vvs.app.util.HeaderMap;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class AliasRegistry {

    private static final Map<String, List<String>> MASTER_ALIASES = Map.ofEntries(
            Map.entry("Visit Date", List.of("Visit Date", "Appt Date", "VisitDate")),
            Map.entry("RootApptID", List.of("RootApptID", "Root Appt ID", "ROOT", "Root_ID")),
            Map.entry("Customer", List.of("Customer", "Client", "Customer Name")),
            Map.entry("Phone", List.of("Phone", "PhoneNorm", "Primary Phone")),
            Map.entry("Email", List.of("Email", "EmailLower", "Primary Email")),
            Map.entry("Visit Type", List.of("Visit Type", "Type", "Appt Type")),
            Map.entry("Visit #", List.of("Visit #", "Visit Number", "VisitNum")),
            Map.entry("SO#", List.of("SO#", "SO Number", "Sales Order")),
            Map.entry("Brand", List.of("Brand", "Brand Name")),
            Map.entry("Sales Stage", List.of("Sales Stage", "Stage", "SalesStage")),
            Map.entry("Conversion Status", List.of("Conversion Status", "Conversion", "Conv Status")),
            Map.entry("Custom Order Status", List.of("Custom Order Status", "CO Status", "CustomOrderStatus")),
            Map.entry("Center Stone Order Status", List.of("Center Stone Order Status", "CS Order Status", "CenterStoneStatus")),
            Map.entry("Assigned Rep", List.of("Assigned Rep", "Assigned Sales Rep", "AssignedRep")),
            Map.entry("Assisted Rep", List.of("Assisted Rep", "Assisted Sales Rep", "AssistedRep"))
    );

    private static final Map<String, List<String>> LEDGER_ALIASES = Map.ofEntries(
            Map.entry("PaymentDateTime", List.of("PaymentDateTime", "Payment DateTime", "Payment Date", "Paid At")),
            Map.entry("AmountNet", List.of("AmountNet", "Net", "Net Amount")),
            Map.entry("DocType", List.of("DocType", "Document Type", "Type")),
            Map.entry("DocStatus", List.of("DocStatus", "Status", "Document Status")),
            Map.entry("RootApptID", MASTER_ALIASES.get("RootApptID")),
            Map.entry("SO#", MASTER_ALIASES.get("SO#"))
    );

    private static final Map<String, List<String>> CLIENT_STATUS_ALIASES = Map.ofEntries(
            Map.entry("RootApptID", MASTER_ALIASES.get("RootApptID")),
            Map.entry("Log Date", List.of("Log Date", "Status Date", "LogDate")),
            Map.entry("Status", List.of("Status", "Client Status", "Current Status")),
            Map.entry("Updated At", List.of("Updated At", "Last Updated", "UpdatedAt"))
    );

    private AliasRegistry() {
    }

    public static List<String> masterAliases(String canonical) {
        return MASTER_ALIASES.getOrDefault(canonical, List.of(canonical));
    }

    public static List<String> ledgerAliases(String canonical) {
        return LEDGER_ALIASES.getOrDefault(canonical, List.of(canonical));
    }

    public static List<String> clientStatusAliases(String canonical) {
        return CLIENT_STATUS_ALIASES.getOrDefault(canonical, List.of(canonical));
    }

    public static Optional<String> resolve(HeaderMap headerMap, List<String> aliases) {
        for (String alias : aliases) {
            Optional<String> header = headerMap.findHeader(alias);
            if (header.isPresent()) {
                return header;
            }
        }
        return Optional.empty();
    }

    public static Optional<String> resolveMaster(HeaderMap headerMap, String canonical) {
        return resolve(headerMap, masterAliases(canonical));
    }

    public static Optional<String> resolveLedger(HeaderMap headerMap, String canonical) {
        return resolve(headerMap, ledgerAliases(canonical));
    }

    public static Optional<String> resolveClientStatus(HeaderMap headerMap, String canonical) {
        return resolve(headerMap, clientStatusAliases(canonical));
    }

    public static Map<String, List<String>> masterAliasMap() {
        return Collections.unmodifiableMap(MASTER_ALIASES);
    }

    public static Map<String, List<String>> ledgerAliasMap() {
        return Collections.unmodifiableMap(LEDGER_ALIASES);
    }

    public static Map<String, List<String>> clientStatusAliasMap() {
        return Collections.unmodifiableMap(CLIENT_STATUS_ALIASES);
    }
}

File: src/main/alias-registry.js
import { normalizeHeaderName } from './utils/headerMap.js';

const MASTER_HEADER_ALIASES = {
  VisitDate: ['Visit Date', 'VisitDate', 'Appt Date', 'Appointment Date'],
  RootApptID: ['RootApptID', 'Root Appt ID', 'ROOT', 'Root_ID'],
  Customer: ['Customer', 'Customer Name', 'Client Name'],
  Phone: ['Phone', 'Phone Number', 'Primary Phone'],
  PhoneNorm: ['PhoneNorm', 'Phone Normalized', 'Phone (Normalized)'],
  Email: ['Email', 'Customer Email'],
  EmailLower: ['EmailLower', 'Email (Lower)', 'Email Lower'],
  VisitType: ['Visit Type', 'Appt Type', 'Type'],
  VisitNumber: ['Visit #', 'Visit Number', 'Appt #'],
  SO: ['SO#', 'SO Number', 'Sales Order'],
  Brand: ['Brand'],
  SalesStage: ['Sales Stage', 'Stage'],
  ConversionStatus: ['Conversion Status', 'Conversion'],
  CustomOrderStatus: ['Custom Order Status', 'Custom Status'],
  CenterStoneOrderStatus: ['Center Stone Order Status', 'Center Stone Status'],
  AssignedRep: ['Assigned Rep', 'Primary Rep'],
  AssistedRep: ['Assisted Rep', 'Secondary Rep'],
};

const LEDGER_HEADER_ALIASES = {
  RootApptID: ['RootApptID', 'Root Appt ID', 'ROOT', 'Root_ID'],
  PaymentDateTime: ['PaymentDateTime', 'Payment DateTime', 'Payment Date', 'Paid At'],
  DocType: ['DocType', 'Document Type', 'Type'],
  AmountNet: ['AmountNet', 'Net', 'Net Amount'],
  DocStatus: ['DocStatus', 'Status'],
};

const SNAPSHOT_ALIASES = {
  SnapshotDate: ['Snapshot Date'],
  CapturedAt: ['Captured At'],
  RootApptID: ['RootApptID', 'Root Appt ID'],
  Rep: ['Rep', 'Assigned Rep'],
  Role: ['Role'],
  ScopeGroup: ['Scope Group'],
  CustomerName: ['Customer Name', 'Customer'],
  SalesStage: ['Sales Stage', 'Stage'],
  ConversionStatus: ['Conversion Status'],
  CustomOrderStatus: ['Custom Order Status'],
  UpdatedBy: ['Updated By'],
  UpdatedAt: ['Updated At'],
  DaysSinceLastUpdate: ['Days Since Last Update'],
  ClientStatusReportURL: ['Client Status Report URL'],
};

const PER_CLIENT_LOG_HEADERS = {
  LogDate: ['Log Date'],
  SalesStage: ['Sales Stage'],
  ConversionStatus: ['Conversion Status'],
  CustomOrderStatus: ['Custom Order Status'],
  CenterStoneOrderStatus: ['Center Stone Order Status'],
  NextSteps: ['Next Steps'],
  DeadlineType: ['Deadline Type'],
  DeadlineDate: ['Deadline Date'],
  MoveCount: ['Move Count'],
  AssistedRep: ['Assisted Rep'],
  UpdatedBy: ['Updated By'],
  UpdatedAt: ['Updated At'],
};

export const APPOINTMENT_SUMMARY_HEADERS = [
  'Visit Date',
  'RootApptID',
  'Customer',
  'Phone',
  'Email',
  'Visit Type',
  'Visit #',
  'SO#',
  'Brand',
  'Sales Stage',
  'Conversion Status',
  'Custom Order Status',
  'Center Stone Order Status',
  'Assigned Rep',
  'Assisted Rep',
];

function resolveHeader(headers, aliases, canonical) {
  const candidates = aliases[canonical] || [canonical];
  const normalizedHeaders = headers.map((header) => normalizeHeaderName(header));
  for (const candidate of candidates) {
    const normalized = normalizeHeaderName(candidate);
    const index = normalizedHeaders.indexOf(normalized);
    if (index !== -1) {
      return headers[index];
    }
  }
  return null;
}

export function selectFirstMatchingHeader(headers, canonical, collection = MASTER_HEADER_ALIASES) {
  return resolveHeader(headers, collection, canonical);
}

export function getAliasSet(collectionName) {
  switch (collectionName) {
    case 'master':
      return MASTER_HEADER_ALIASES;
    case 'ledger':
      return LEDGER_HEADER_ALIASES;
    case 'snapshot':
      return SNAPSHOT_ALIASES;
    case 'perClient':
      return PER_CLIENT_LOG_HEADERS;
    default:
      return {};
  }
}

export {
  MASTER_HEADER_ALIASES,
  LEDGER_HEADER_ALIASES,
  SNAPSHOT_ALIASES,
  PER_CLIENT_LOG_HEADERS,
};
